### ISSUE TITLE
training run: tuning ichigo-quantizer

### ISSUE THREAD

**Issue by hahuyhoang411** [janhq/ichigo#144]
# Goal
Tuning to find the best params for Ichigo Quantizer

# Hypothesis
- Current default codebook (512) is too small for multilingual, and also training from scratch needs so much data to converge
- Training on mixed dataset (Eng-Vie) maintains the performance of original WhisperVQ

# Task:
- [x] Adding code to init codebook embedding from [WhisperVQ checkpoint](https://huggingface.co/WhisperSpeech/WhisperSpeech/blob/main/whisper-vq-stoks-v3-7lang.model)
- [x] Fix training loss stuck
- [x] Adding code to preprocess multi training datasets ([Filtered LibriTTS-R](https://huggingface.co/datasets/parler-tts/libritts_r_filtered) + Viet-Bud500)
- [x] Adding inference quantizer and fill [Google Sheet](https://docs.google.com/spreadsheets/d/1NR6BVIC8djxKxMpkHNdkEUaOgA8a6OxO_MhfnpTOUUA/edit?usp=sharing)
- [x] Packing short audio (~3s) data (up to 30s)
- [x] Verify KL loss
- [x] Add WER for Vietnamese 
- [ ] Scale training on large dataset (ViVoice, MLS, etc.)

# Training Result:
The training result here: https://github.com/janhq/WhisperSpeech/issues/2

**Comment by tuanlda78202** [comment]
Fix inference bug, remove `input_toks` and apply autoregressive `whisper.DecodingOptions()`

WER comparison
Dataset | Language | Trained Quantizer | Pho Whisper | Whisper Medium
-- | -- | ------ | ------| ------------ 
Bud500 | Vi| 0.22 | 0.08 | 0.62
LibriTTS-R | En | 1.90 | 0.46 | 0.12

**Comment by tikikun** [comment]
# Errors in data samplings
- We should treat clean and finished data like libri separately, there is no need for excess concatenation like below.
- Concatenation should only happen in low resource languages.
- Reduce the number of training epochs.

![image](https://github.com/user-attachments/assets/5f704d2f-1003-4a20-8d73-6fd68707d1a7)

**Comment by tikikun** [comment]
## Next test run:
- 2 phases:
+ Phase 1: KL Loss + CE Loss
+ Phase 2: CE Loss
+ 10 epochs for each phase

## What to validate
- Is the English degradation is due to wrong data samplings (above error?)
- Are we overtrain with current 100 epochs?

Only after validated the above points we can move forwards with next steps

cc @tuanlda78202 TBD today

**Comment by tuanlda78202** [comment]
# Problem
1. Are we overtraining with the current 100 epochs for both training phases?
2. Is the English degradation due to concatenating audio and the sampling distribution being skewed to the Vietnamese language?

# Solution
1. Reduce the training epochs for both training phases.
2. Train on the original English (non-concatenate) dataset and balance the sampling distribution on a per-batch basis to regularize the model.

# Implementation

1. Train two phases, each phase with 10 epochs. Phase 1 turns off KL loss and weights the dataset distribution to `0.5` for concatenating Bud500 and non-concatenating LibriTTS-R.

Name | Epoch | KL Loss | Val Loss | Val Acc
-- | -- | -- | -- | --
Phase 1  | 10 | On  | 15.56 | 0.84
Phase 2  | 10 | Off | 0.21  | 0.94

2. Inference checkpoints are saved after 10 epochs for both phases on ViVoice (100 samples) and LibriTTS-R.

| <img width="595" alt="Image" src="https://github.com/user-attachments/assets/442d73aa-9341-4183-bd7b-a268a7a9c845" /> | 
|:--:| 
| ViVoice first 100 samples (Vietnamese)|

| <img width="598" alt="Image" src="https://github.com/user-attachments/assets/f2440464-c3c3-4505-a588-b156159fd322" /> | 
|:--:| 
| LibriTTS-R (English) |

# Conclusion

## Observation
* Training for 10 epochs improves performance, but extending to 100 epochs degrades it, although the first 100-epoch performance is better than 10-epoch performance.

## Solution
* Train first 100 epochs with KL loss
* Fine-tune for 10 epochs without KL loss.

**Comment by tuanlda78202** [comment]
# Problem 
* Training model on low-quality datasets (Bud500) lead to poor performance 

# Solution
* Training model on high-quality datasets (viVoice)

# Results 
1. Phase 1 (with KL loss): Training on viVoice (868k samples) and LibriTTS-R (112k samples), not use concat dataset, early stop at 10 epoch, result on val dataset (10k samples)

Best epoch | Accuracy | Loss
------------------------------ | ------------------------------| ------------------------------
29 | 0.89 | 14.59

Testing: Ongoing

2. Phase 2 (without KL loss): tbc

**Comment by tuanlda78202** [comment]
## Phase 2 (without KL loss)

### Phase 1 best epoch (29), Phase 2 10e

1. viVoice

Model Name | Phase | Epoch | Dataset train | Dataset test | Language Test | Test samples |WER
----------------- | ----------------- |----------------- |----------------- |----------------- |----------------- | ----------------- |----------------- |
**Ichigo Quantizer** | **2** | **2** | **viVoice + LibriTTS-R** | **viVoice** | **Vi** | **10000** | **0.20**
PhoWhisper Large | - | - | - | viVoice | Vi | 10000 | 0.24
Whisper Medium | - | - | - | viVoice | Vi | 10000 | 0.18
**Ichigo Quantizer** | **2** | **9** | **viVoice + LibriTTS-R** | **LibriTTS-R** | **Vi** | **1000** | **0.11**
PhoWhisper Large | - | - | - | viVoice | Vi | 1000 | 0.23
Whisper Medium | - | - | - | viVoice | Vi | 1000 | 0.18

2. LibriTTS-R

Model Name | Phase | Epoch | Dataset train | Dataset test | Language Test | Test samples |WER
----------------- | ----------------- |----------------- |----------------- |----------------- |----------------- | ----------------- |----------------- |
**Ichigo Quantizer** | **2** | **2** | **viVoice + LibriTTS-R** | **LibriTTS-R** | **En** | **4689** | **1.23**
PhoWhisper Large | - | - | - | LibriTTS-R | En | 4689 | 0.47
Whisper Medium | - | - | - | LibriTTS-R | En | 4689 | 0.12
**Ichigo Quantizer** | **2** | **9** | **viVoice + LibriTTS-R** | **LibriTTS-R** | **En** | **1000** | **0.70**
PhoWhisper Large | - | - | - | LibriTTS-R | En | 1000 | 0.59
Whisper Medium | - | - | - | LibriTTS-R | En | 1000 | 0.12

### Phase 1 full epoch (100), Phase 2 10e [tbc]

**Comment by tuanlda78202** [comment]
# Phase 1 full epoch (100), Phase 2 10e [Ongoing]
## Phase 1 

1. viVoice

Model Name | Phase | Epoch | Dataset train | Dataset test | Language Test | Test samples | WER | WER (%)
----------------- | ----------------- |----------------- |----------------- |----------------- |----------------- | ----------------- | ----------------- | ----------------- |
**Ichigo Quantizer** | **1** | **29** | **viVoice + LibriTTS-R** | **viVoice** | **Vi** | **10000** | **0.21** | **21**
PhoWhisper Large | - | - | - | viVoice | Vi | 10000 | 0.23 | 23
Whisper Medium | - | - | - | viVoice | Vi | 10000 | 0.18 | 18
**Ichigo Quantizer** | **1** | **62** | **viVoice + LibriTTS-R** | **viVoice** | **Vi** | **1000** | **0.18** | **18**
PhoWhisper Large | - | - | - | viVoice | Vi | 1000 | 0.23 | 23
Whisper Medium | - | - | - | viVoice | Vi | 1000 | 0.18 | 18

2. LibriTTS-R

Model Name | Phase | Epoch | Dataset train | Dataset test | Language Test | Test samples | WER | WER (%)
----------------- | ----------------- |----------------- |----------------- |----------------- |----------------- | ----------------- | ----------------- | ----------------- |
**Ichigo Quantizer** | **1** | **29** | **viVoice + LibriTTS-R** | **LibriTTS-R** | **En** | **4689** | **0.13** | **13**
PhoWhisper Large | - | - | - | LibriTTS-R | En | 4689 | 0.47 | 47
Whisper Medium | - | - | - | LibriTTS-R | En | 4689 | 0.12 | 12
**Ichigo Quantizer** | **1** | **62** | **viVoice + LibriTTS-R** | **LibriTTS-R** | **En** | **1000** | **0.13** | **13**
PhoWhisper Large | - | - | - | LibriTTS-R | En | 1000 | 0.59 | 59
Whisper Medium | - | - | - | LibriTTS-R | En | 1000 | 0.12 | 12

**Comment by bachvudinh** [comment]
# Bug: Incorrect Mask Generation After Audio Padding

## Description
During code review, we discovered that the mask generation after audio padding is incorrectly implemented. The current code creates masks with all 1s for padded audio, causing the model to train on padding tokens instead of masking them out.

```python
# Current problematic code
concatenated_audio = self.pad_audio(concatenated_audio)
mask = torch.zeros(30 * 16000 // 320, dtype=torch.bool)
audio_frames = min(len(concatenated_audio), self.max_audio_length) // 320
mask[:audio_frames] = 1  # Bug: This includes padding tokens
```

This affects the forward pass:
```python
# Impact on training
if self.training and self.config.mask_embs and mask is not None:
    x[~mask] = project_out(self.rq.layers[0]._codebook.embed[0, self.vq_codes])
```

## Impact
- Model is training on padding tokens instead of masking them
- Training efficiency and model performance potentially compromised

## Priority
High - affects model training quality

## Next Steps
- [ ] Fix mask generation to properly exclude padding tokens
- [ ] Add validation test for mask generation
- [ ] Review impact on currently trained models

**Comment by tuanlda78202** [comment]
# [Merge WhisperVQ codebooks] IchigoWhisper Phase 1 50e, Phase 2 5e

1. viVoice

Model Name | Phase | Merge | Dataset train | Dataset test | Language Test | Test samples | WER (%)
----------------- | ----------------- |----------------- |----------------- |----------------- |----------------- | ----------------- | ----------------- |
**IchigoWhisper** | **2** | No | **viVoice + LibriTTS-R** | **viVoice** | **Vi** | **1000** | **14.46**
**IchigoWhisper** | **2** | Yes | **viVoice + LibriTTS-R** | **viVoice** | **Vi** | **1000** | **11.36**
PhoWhisper Large 1.55B | - | - | - | viVoice | Vi | 1000 | 23.08
Whisper Medium 0.76B | - | - | - | viVoice | Vi | 1000 | 18.64

2. LibriTTS-R

Model Name | Phase | Merge | Dataset train | Dataset test | Language Test | Test samples | WER (%)
----------------- | ----------------- |----------------- |----------------- |----------------- |----------------- | ----------------- | ----------------- |
**IchigoWhisper** | **2** | No | **viVoice + LibriTTS-R** | **LibriTTS-R** | **En** | **1000** | **16.78**
**IchigoWhisper** | **2** | Yes | **viVoice + LibriTTS-R** | **LibriTTS-R** | **En** | **1000** | **13.01**
PhoWhisper Large 1.55B | - | - | - | LibriTTS-R | En | 1000 | 59.72
Whisper Medium 0.76B | - | - | - | LibriTTS-R | En | 1000 | 12.99

**Comment by hahuyhoang411** [comment]
I updated the [Readme](https://huggingface.co/homebrewltd/Ichigo-whisper-v0.1) for this run. Several information I needed to fill in 
- [ ] hours training in English + Vietnamese
- [ ] Training specs (LR, hardware, etc)
- [ ] example load + infer code

@tuanlda78202

**Comment by tuanlda78202** [comment]
# [Full] IchigoWhisper (non-cut PAD) Phase 1 50e, Phase 2 20e

## viVoice (10000)

Model Name | Phase | Codebook Size | Base VQ |Dataset train | Dataset test | Language Test | Test samples | WER
----------------- | ----------------- |----------------- |-----------------|----------------- |----------------- |----------------- | ----------------- | ----------------- |
**IchigoWhisper** | **2** | 2561 | `v3-7lang`|**viVoice + LibriTTS-R** | **viVoice** | **Vi** | **10000** | **11.68**
PhoWhisper Large 1.55B | - | - | - | - |  viVoice | Vi | 10000 | 24.00
Whisper Medium 0.76B | - | - | - | - |  viVoice | Vi | 10000 | 18.30

## LibriTTS-R (4689)

Model Name | Phase | Codebook Size | Base VQ |Dataset train | Dataset test | Language Test | Test samples | WER
----------------- | ----------------- |----------------- |-----------------|----------------- |----------------- |----------------- | ----------------- | ----------------- |
**IchigoWhisper** | **2** | 2561 | `v3-7lang` |**viVoice + LibriTTS-R** | **LibriTTS-R** | **En** | **4689** | **11.89**
PhoWhisper Large 1.55B | - | - | - | - | LibriTTS-R | En | 4689 | 47.52
Whisper Medium 0.76B | - | - | - | - | LibriTTS-R | En | 4689 | 13.06

**Comment by tuanlda78202** [comment]
# Training Spec

1. Training Hours
a. Phase 1: 75 hours (50 epochs)
b. Phase 2: 29 hours (20 epochs)

2. Specifications
a. GPUs: 8 × A6000
b. Batch Size: 42 per GPU
c. Learning Rate: 1e-3 (Optimizer: AdamW)
d. Scheduler: Linear warm-up for 500 steps, Cosine decay later
f. Weight Decay: 0.001
e. Checkpoint Init: Whisper VQ Large v3 (7 languages), duplication noise ×4, VQ code size 2048
f. Input Padding: Audio padded to 30 seconds (compatible with Whisper)

cc @hahuyhoang411

**Comment by tuanlda78202** [comment]
## First 1000 Test samples

### viVoice (1000)

Model Name | Phase | Epoch | Codebook Size | Base VQ |Dataset train | Dataset test | Language Test | Test samples | WER
----------------- | ----------------- | ----------------- |----------------- |-----------------|----------------- |----------------- |----------------- | ----------------- | ----------------- |
**Ichigo Whisper** | 2 | 10 | 2049 | `v3-7lang`| viVoice + LibriTTS-R | viVoice | Vi | 1000 | 11.10 
**Ichigo Whisper** | 2 | 15 | 2049 | `v3-7lang`| viVoice + LibriTTS-R | viVoice | Vi | 1000 | 11.00 
**Ichigo Whisper** | 2 | 20 | 2049 | `v3-7lang`| viVoice + LibriTTS-R | viVoice | Vi | 1000 | 10.83 
PhoWhisper Large 1.55B | - | - | - | - | - |  viVoice | Vi | 1000 | 23.08
Whisper Medium 0.76B | - | - | - | - | - |  viVoice | Vi | 1000 | 19.36

### LibriTTS-R (1000)

Model Name | Phase | Epoch | Codebook Size | Base VQ |Dataset train | Dataset test | Language Test | Test samples | WER
----------------- | ----------------- | ---------------- |----------------- |-----------------|----------------- |----------------- |----------------- | ----------------- | ----------------- |
**Ichigo Whisper** | 2 | 10 | 2049 | `v3-7lang` | viVoice + LibriTTS-R | LibriTTS-R | En | 1000 | 23.35
**Ichigo Whisper** | 2 | 15 | 2049 | `v3-7lang` | viVoice + LibriTTS-R | LibriTTS-R | En | 1000 | 17.72
**Ichigo Whisper** | 2 | 20 | 2049 | `v3-7lang` | viVoice + LibriTTS-R | LibriTTS-R | En | 1000 | 15.97
PhoWhisper Large 1.55B | - | - | - |- | - | LibriTTS-R | En | 1000 | 59.72
Whisper Medium 0.76B | - | - | - | - | - | LibriTTS-R | En | 1000 | 12.99
